src/transform/collect_free_vars.rs

```diff
@@ -0,0 +1,80 @@
+use std::collections::HashSet;
+
+use crate::destruct::{Destruct, Match};
+use crate::term::{SharedTerm, TermType};
+use crate::{
+    identifier::Ident,
+    term::{RichTerm, Term},
+};
+
+// `collect_free_vars` is transformation pass that caches a terms free variables.
+// This is used for optimizing memory consumption
+//
+// `free_vars` is the set of free variables in this term.
+//
+// `fields_free_vars` is a mapping between position of a field in a RecRecord`,
+//  and the free variables used in the body of the field.
+pub fn collect_free_vars(
+    rt: &mut RichTerm,
+    parent_rec_record: TermType,
+    free_vars: &mut HashSet<Ident>,
+    fields_free_vars: &mut Vec<HashSet<Ident>>,
+) {
+    if parent_rec_record == TermType::RecRecord {
+        fields_free_vars.push(free_vars.clone());
```

Comment:
Doesn't this mean that you keep accumulating the free variables of the previous fields? Like in:

```
{x = y, foo = 1}
```

Assuming you iterate on `foo` in second, I think you will have `y` as free variable of `foo`, unduly. I suspect you need to set `*free_vars = HashSet::new()` here. It does mean that to compute the free vars of a recursive record, you need to merge all the free vars into one again. You could have yet another variable to keep track of the full set of free variables, but I think it's premature optimization, and it's fine to have to merge hashsets on a recursive record.